### ISSUE TITLE
[BUG] Call to key vault may hangs forever  

### ISSUE THREAD

**Issue by werder06** [Azure/azure-sdk-for-java#28364]
I use azure-security-keyvault-secrets v.4.3.6 on java17 application to communicate with Azure Key Vault. 
Application receives external call, makes some processing, part of which is call to Key Vault, and returns result.

Code for key vault communication is as simple as: 

```
// create client bean once
public SecretClient secretsClient(...) {
    return new SecretClientBuilder()
                .vaultUrl(...)
                .credential(...)
                .addPolicy(new TimeoutPolicy(...)) //60 sec
                .httpLogOptions(new HttpLogOptions().setLogLevel(HttpLogDetailLevel.BASIC))
                .buildClient();
    }

...
// make calls for secret 
    public String fetch(String secretName) {
        try {
            return secretClient.getSecret(secretName).getValue();
        } catch (ResourceNotFoundException e) {
               ...
        }
    }
```

For last two months, on different envs (stage, prod, different regions) periodically I've been having this issue:
- Application stops to respond
- Application utilizes 100% CPU 
- In thread dump I can see that all working threads are waiting for key vault response (seems forever, there are no timeout exceptions):

```
"..." #32 prio=5 os_prio=0 cpu=22.22ms elapsed=244.85s tid=0x00007f3ce8061790 nid=0x2a waiting on condition  [0x00007f3cdecfa000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@17.0.2/Native Method)
	- parking to wait for  <0x00000000c14c4a20> (a java.util.concurrent.CountDownLatch$Sync)
	at java.util.concurrent.locks.LockSupport.park(java.base@17.0.2/LockSupport.java:211)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@17.0.2/AbstractQueuedSynchronizer.java:715)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(java.base@17.0.2/AbstractQueuedSynchronizer.java:1047)
	at java.util.concurrent.CountDownLatch.await(java.base@17.0.2/CountDownLatch.java:230)
	at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:87)
	at reactor.core.publisher.Mono.block(Mono.java:1707)
	at com.azure.security.keyvault.secrets.SecretClient.getSecretWithResponse(SecretClient.java:228)
	at com.azure.security.keyvault.secrets.SecretClient.getSecret(SecretClient.java:198)

```
...while reactor threads seems to be very busy on some reactor stuff which probably costs 100% of CPU [reactor_thread_dump.txt](https://github.com/Azure/azure-sdk-for-java/files/8507205/reactor_thread_dump.txt)

Just pods restart didn't help. After some attempts and experiments I was able to overcome issue by follow list of steps:
- In portal turn off/ turn on firewall for key vault
- Wait for couple of minutes
- Restart pods 

I've done this several times, so even if steps seems to be messy, that really works for me. 

There are probably some 'accessibility issues' with key vault. But seems that client library handles it in a not best way possible: 
- Timeout didn't fail response 
- After 'accessibility issues' with key vault are gone, there still has to be pod restart to reload working threads state 

Any ideas or suggestions what could be wrong or how to avoid issue in future are very welcome.

**Comment by joshfree** [comment]
Hi @andrey-kudryavtsev-parsable, thank you for reaching out to us via GitHub.  We'll have the right area experts follow up with you shortly.  @vcolin7 @anuchandy could you please take a look?

**Comment by vcolin7** [comment]
Sure thing @joshfree, will look into this first thing tomorrow :)

**Comment by vcolin7** [comment]
Hi @andrey-kudryavtsev-parsable, @anuchandy and I spent some time looking into this late last week. We think we found out a likely cause for the high CPU usage when a service call hangs for a long time and are working on coming up with a fix.
I just have a couple clarifying questions about your scenario:
1. **What type of credential object is provided to the `SecretClientBuilder.credential()`?**
2. **Are you retrieving secrets in multiple threads simultaneously using the same `SecretClient`? If so, how are you achieving this?** A code sample would be very welcome.
3. **Are you performing other intensive tasks that could be taking up most of your main thread pool?**
4. **Next time you see this happen, could you restart the pod without touching the Key Vault firewall?** We'd like to collect logs as to what is making the Key Vault call hang. To do so, you can enable the environment variable `AZURE_LOG_LEVEL` to `DEBUG` before restarting. [Here](https://docs.microsoft.com/azure/developer/java/sdk/logging-overview) is more info on how to enable logging in the Azure SDK for Java.

Additionally, we'd like to collect some more sensitive information to figure out what's happening on the service side as well. There are a couple channels you can choose from for this:

- Send us an email to azjavasdkhelp@microsoft.com with the subject `Follow-up from GitHub - [BUG] Call to key vault may hangs forever #28364`.

**OR**

- If you have an Azure support plan, you can open an azure support ticket in the Azure Portal by going to your Key Vault resource and pressing "New Support Request" on the bottom left. [Here's more information](https://github.com/Azure/azure-sdk-for-java/blob/main/SUPPORT.md#azure-support-tickets) about this process. **If you decide to do this, make sure to mention this GitHub issue in the description.**

Let me know which option you selected and we'll follow up from there :)

**Comment by andrey-kudryavtsev-parsable** [comment]
Hi, @vcolin7 

> What type of credential object is provided to the SecretClientBuilder.credential()?

ClientSecretCredential implementation which is created like: 

```
   public TokenCredential tokenCredential(AzureProperties azureProperties) {
        return new ClientSecretCredentialBuilder()
                .clientId(azureProperties.getAuth().getClientId())
                .clientSecret(azureProperties.getAuth().getClientSecret())
                .tenantId(azureProperties.getAuth().getTenantId())
                .build();
    }
```

> Are you retrieving secrets in multiple threads simultaneously using the same SecretClient? If so, how are you achieving this? A code sample would be very welcome.

The answer is yes, we use SecretClient instance in multiple threads environment. SecretClient instance is created as a Spring @Bean, set up into Spring @Component as a property. Other parts of code calls that component's method: 
```  public String fetch(String secretName) {
        try {
            return secretClient.getSecret(secretName).getValue();
        } catch (ResourceNotFoundException e) {
               ...
        }
    }
```

> Are you performing other intensive tasks that could be taking up most of your main thread pool?
I don't think so. Application logic is simple and is not CPU-intensive (mostly IO operations). It seems that only thats 'reactor' threads are in running state at the most of the time

> Next time you see this happen, could you restart the pod without touching the Key Vault firewall? We'd like to collect logs as to what is making the Key Vault call hang. To do so, you can enable the environment variable AZURE_LOG_LEVEL to DEBUG before restarting. [Here](https://docs.microsoft.com/azure/developer/java/sdk/logging-overview) is more info on how to enable logging in the Azure SDK for Java.

Nice option, thank you. Already missed one good opportunity to do that.

> Additionally, we'd like to collect some more sensitive information to figure out what's happening on the service side as well. There are a couple channels you can choose from for this:...

Will reach you by email if you don't mind. Thank you.  

btw, have an opportunity to update client lib to 4.3.7. Does it make any sense? Maybe reactor or/and netty update?

**Comment by vcolin7** [comment]
@andrey-kudryavtsev-parsable I think there was a mixup with the email address you tried to reach: you said you emailed [azsdkjava@microsoft.com](mailto:azsdkjava@microsoft.com), can you try emailing [azjavasdkhelp@microsoft.com](mailto:azjavasdkhelp@microsoft.com) instead?

Also, in the meantime could you try passing an `ExecutorService` to the `ClientSecretCredentialBuilder`? This will ensure token retrieval happens in a thread pool other than the main one:

```java
ClientSecretCredential secretCredential = new ClientSecretCredentialBuilder()
    .clientSecret(...)
    .clientId(...)
    .tenantId(...)
    .executorService(executorService)
    .build();
```

Just remember to ensure the `ExecutorService` is shut down before the application is terminated (if ever). The following is a simple example with a try/finally block:

```java
ExecutorService executorService = Executors.newCachedThreadPool();

try {
    ClientSecretCredential secretCredential = new ClientSecretCredentialBuilder()
        .clientSecret(...)
        .clientId(...)
        .tenantId(...)
        .executorService(executorService)
        .build();
    SecretClient secretClient = new SecretClientBuilder()
        .vaultUrl(...)
        .credential(...)
        .buildClient();

    // App code which utilizes client.
} finally {
    // Shutdown before application terminates.
    if (executorService != null && !executorService.isTerminated()) {
        executorService.shutdown();
    }
}
```

**Comment by andrey-kudryavtsev-parsable** [comment]
@vcolin7 I sent it to azjavasdkhelp@microsoft.com, but in response it's like: 

![email](https://user-images.githubusercontent.com/1666226/166649130-df05684e-da6a-4dc1-8cdb-9498b8068179.png)


redirect or something

**Comment by vcolin7** [comment]
I see, thanks for letting me know, I will pass it along and find a way to securely get in touch. By the way, were you able to try out my suggestion involving `ExecutorService`?

**Comment by andrey-kudryavtsev-parsable** [comment]
@vcolin7 I tried suggestion involving `ExecutorService` some time ago, recently service got into same state.

The only difference I see it's a new thread from new pool in stack trace: 

```
"token-retrieval-0" #129 prio=5 os_prio=0 cpu=8.89ms elapsed=959.38s tid=0x00007f0fb4aa02a0 nid=0x8c waiting on condition  [0x00007f0f4c9a7000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@17.0.2/Native Method)
	- parking to wait for  <0x00000000c1970870> (a java.util.concurrent.CountDownLatch$Sync)
	at java.util.concurrent.locks.LockSupport.park(java.base@17.0.2/LockSupport.java:211)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@17.0.2/AbstractQueuedSynchronizer.java:715)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(java.base@17.0.2/AbstractQueuedSynchronizer.java:1047)
	at java.util.concurrent.CountDownLatch.await(java.base@17.0.2/CountDownLatch.java:230)
	at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:87)
	at reactor.core.publisher.Mono.block(Mono.java:1707)
	at com.azure.identity.implementation.HttpPipelineAdapter.send(HttpPipelineAdapter.java:58)
	at com.microsoft.aad.msal4j.HttpHelper.executeHttpRequestWithRetries(HttpHelper.java:96)
	at com.microsoft.aad.msal4j.HttpHelper.executeHttpRequest(HttpHelper.java:49)
	at com.microsoft.aad.msal4j.AadInstanceDiscoveryProvider.executeRequest(AadInstanceDiscoveryProvider.java:232)
	at com.microsoft.aad.msal4j.AadInstanceDiscoveryProvider.sendInstanceDiscoveryRequest(AadInstanceDiscoveryProvider.java:187)
	at com.microsoft.aad.msal4j.AadInstanceDiscoveryProvider.doInstanceDiscoveryAndCache(AadInstanceDiscoveryProvider.java:285)
	at com.microsoft.aad.msal4j.AadInstanceDiscoveryProvider.getMetadataEntry(AadInstanceDiscoveryProvider.java:57)
	at com.microsoft.aad.msal4j.AuthenticationResultSupplier.getAuthorityWithPrefNetworkHost(AuthenticationResultSupplier.java:39)
	at com.microsoft.aad.msal4j.AcquireTokenSilentSupplier.execute(AcquireTokenSilentSupplier.java:23)
	at com.microsoft.aad.msal4j.AuthenticationResultSupplier.get(AuthenticationResultSupplier.java:69)
	at com.microsoft.aad.msal4j.AuthenticationResultSupplier.get(AuthenticationResultSupplier.java:18)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(java.base@17.0.2/CompletableFuture.java:1768)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@17.0.2/ThreadPoolExecutor.java:1136)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@17.0.2/ThreadPoolExecutor.java:635)
	at java.lang.Thread.run(java.base@17.0.2/Thread.java:833)

```

**Comment by vcolin7** [comment]
Thanks for sharing this @andrey-kudryavtsev-parsable, I'll pass it along to the team. By the way, I looked into the email issue and I think you should now be able to reach us at the address I gave you before (azjavasdkhelp@microsoft.com).

**Comment by gukoff** [comment]
@werder06 

My team sees the same issue as you, except when deleting secrets from parallel threads. Deleting seems to be hanging forever, and the app consumes 100% CPU.

Could you please answer a few questions to help us understand the next troubleshooting steps:
- Did you manage to resolve this bug?
- Do you use opentelemetry in the app?
- Have you tried to use `SecretAsyncClient` instead of `SecretClient`?
- Have you tried adding HttpClient-level timeouts as follows:
```java
var httpClientOptions = new HttpClientOptions();
httpClientOptions.setConnectTimeout(Duration.ofSeconds(1));
httpClientOptions.setReadTimeout(Duration.ofSeconds(1));
httpClientOptions.setResponseTimeout(Duration.ofSeconds(1));
httpClientOptions.setWriteTimeout(Duration.ofSeconds(1));
httpClientOptions.setConnectionIdleTimeout(Duration.ofSeconds(1));

var client = new SecretClientBuilder()
    .vaultUrl(VaultUrl)
    .addPolicy(new TimeoutPolicy(Duration.ofSeconds(100)))
    .credential(credential)
    .clientOptions(httpClientOptions)
    .buildClient();
```

---

**Our stacktraces**


In the thread dump we're seeing 5 "boundedElastic-N" threads with `keyvault.secrets.SecretClient.beginDeleteSecret` in the stacktrace. 4 consume little CPU, and 1 consumes a lot. Their stack traces are a little different.


4 threads are WAITING and look like this:

```
"boundedElastic-12" #105 daemon prio=5 os_prio=0 cpu=164.26ms elapsed=4943.77s allocated=34273K defined_classes=75 tid=0x00007fa8cf06e800 nid=0x7c waiting on condition  [0x00007fa82b897000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@11.0.16/Native Method)
	- parking to wait for  <0x00000000e051c530> (a java.util.concurrent.CountDownLatch$Sync)
	at java.util.concurrent.locks.LockSupport.park(java.base@11.0.16/Unknown Source)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(java.base@11.0.16/Unknown Source)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(java.base@11.0.16/Unknown Source)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(java.base@11.0.16/Unknown Source)
	at java.util.concurrent.CountDownLatch.await(java.base@11.0.16/Unknown Source)
	at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:87)
	at reactor.core.publisher.Mono.block(Mono.java:1706)
	at com.azure.core.util.polling.PollerFlux.lambda$new$1(PollerFlux.java:289)
	at com.azure.core.util.polling.PollerFlux$$Lambda$2388/0x0000000841318c40.apply(Unknown Source)
	at com.azure.core.util.polling.DefaultSyncPoller.<init>(DefaultSyncPoller.java:71)
	at com.azure.core.util.polling.PollerFlux.getSyncPoller(PollerFlux.java:467)
	at com.azure.security.keyvault.secrets.SecretClient.beginDeleteSecret(SecretClient.java:337)
```

And one very CPU-active thread is RUNNABLE and looks like this:

```
"boundedElastic-11" #104 daemon prio=5 os_prio=0 cpu=2387754.05ms elapsed=4943.77s allocated=353G defined_classes=108 tid=0x00007fa8ce631000 nid=0x7b runnable  [0x00007fa82b996000]
   java.lang.Thread.State: RUNNABLE
	at reactor.core.publisher.FluxLiftFuseable.subscribeOrReturn(FluxLiftFuseable.java:58)
	at reactor.core.publisher.Flux.subscribe(Flux.java:8454)
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.drain(FluxConcatMap.java:449)
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.innerComplete(FluxConcatMap.java:296)
	at reactor.core.publisher.FluxConcatMap$ConcatMapInner.onComplete(FluxConcatMap.java:885)
	at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2400)
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.request(Operators.java:2158)
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.request(FluxConcatMap.java:330)
	at reactor.core.publisher.Operators$DeferredSubscription.request(Operators.java:1680)
	at reactor.core.publisher.FluxRepeatWhen$RepeatWhenMainSubscriber.onComplete(FluxRepeatWhen.java:164)
	at io.opentelemetry.javaagent.shaded.instrumentation.reactor.TracingSubscriber$$Lambda$1617/0x0000000840fb3440.run(Unknown Source)
	at io.opentelemetry.javaagent.shaded.instrumentation.reactor.TracingSubscriber.withActiveSpan(TracingSubscriber.java:79)
	at io.opentelemetry.javaagent.shaded.instrumentation.reactor.TracingSubscriber.onComplete(TracingSubscriber.java:68)
	at reactor.core.publisher.Operators.complete(Operators.java:137)
	at reactor.core.publisher.MonoEmpty.subscribe(MonoEmpty.java:46)
	at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52)
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64)
	at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52)
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64)
	at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4399)
	at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.subscribeNext(MonoIgnoreThen.java:255)
	at reactor.core.publisher.MonoIgnoreThen.subscribe(MonoIgnoreThen.java:51)
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64)
	at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52)
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64)
	at reactor.core.publisher.MonoDeferContextual.subscribe(MonoDeferContextual.java:55)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4399)
	at reactor.core.publisher.Mono.block(Mono.java:1705)
	at com.azure.core.util.polling.PollerFlux.lambda$new$1(PollerFlux.java:289)
	at com.azure.core.util.polling.PollerFlux$$Lambda$2388/0x0000000841318c40.apply(Unknown Source)
	at com.azure.core.util.polling.DefaultSyncPoller.<init>(DefaultSyncPoller.java:71)
	at com.azure.core.util.polling.PollerFlux.getSyncPoller(PollerFlux.java:467)
	at com.azure.security.keyvault.secrets.SecretClient.beginDeleteSecret(SecretClient.java:337)
```

**Comment by bennydean** [comment]
Hello @gukoff , I am facing this exact issue. While running a java spring application in a kubernetes pod, the getValue call just hangs and the program locks up. I noticed the pod is getting CPU throttled but there is very little diagnostic information available to try and troubleshoot. Any updates on a solution?

I have tried the following with the standard client and async client with no luck:

```
HttpClientOptions httpClientOptions = new HttpClientOptions();
httpClientOptions.setConnectTimeout(Duration.ofSeconds(1));
httpClientOptions.setReadTimeout(Duration.ofSeconds(1));
httpClientOptions.setResponseTimeout(Duration.ofSeconds(1));
httpClientOptions.setWriteTimeout(Duration.ofSeconds(1));
httpClientOptions.setConnectionIdleTimeout(Duration.ofSeconds(1));

LOGGER.info("Initializing Keyvault Client...");

secretClient = new SecretClientBuilder()
		.vaultUrl(properties.getKeyVaultUri())
		.addPolicy(new TimeoutPolicy(Duration.ofSeconds(100)))
		.credential(
				new ClientSecretCredentialBuilder()
						.clientSecret(properties.getKeyVaultClientSecret())
						.clientId(properties.getKeyVaultClientId())
						.tenantId(properties.getKeyVaultTenantid())
						.build())
		.clientOptions(httpClientOptions)
		.buildClient();

secret = secretClient.getSecret(secret).getValue();

```

**Comment by bennydean** [comment]
After trying all recommendations here including trying newer package versions, the solution for me was to downgrade the azure-security-keyvault-secrets dependency to version 4.2.6

It is now running successfully for me.

**Comment by gukoff** [comment]
@bennydean could you please share more information to help us troubleshoot this?

- Do you have a way to consistently reproduce the issue? Maybe you could create a repository with reproduction steps? We also run code on k8s, but for us the issue is transient, and we've only caught it happen during secret deletion once a month or so.
- Could you attach the stack dump of the hanging application? You can get it [with jstack](https://www.baeldung.com/java-thread-dump).
- Do you use opentelemetry in the app?
- Do you use k8s on AKS?

**Comment by gukoff** [comment]
What is also interesting, we have 4 active CPU-consuming threads in total:
```
"boundedElastic-11" #104 daemon prio=5 os_prio=0 cpu=2387754.05ms elapsed=4943.77s
"reactor-http-epoll-1" #69 daemon prio=5 os_prio=0 cpu=2389104.61ms elapsed=4956.52s
"reactor-http-epoll-2" #70 daemon prio=5 os_prio=0 cpu=2388567.48ms elapsed=4956.52s
"reactor-http-epoll-4" #72 daemon prio=5 os_prio=0 cpu=2388859.10ms elapsed=4956.52s
```

They all, and only they have this line in their stacktraces:

```
FluxRepeatWhen$RepeatWhenMainSubscriber.onComplete(FluxRepeatWhen.java:164)
```

Could it cause infinite looping through monos?

Full stack trace of one of the threads: https://pastebin.com/raw/JS8pA6D9

**Comment by bennydean** [comment]
@gukoff 

- Yes, this happens every time I deploy the code so it is reproduceable. I can't give the code because its a company repo  but I can give you all package / environment configurations. Just let me know what you need to recreate the envrionment.
-  I won't be able to give stack dump with jstack because its unavailable in our k8s environment. 
- We are not using opentelemetry
- We run a k8s cluster in our data center, not on Azure cloud. 
- I have not checked newer versions of 1.1x, so I can't say what the latest version is but I have checked a version in 1.2x and 1.3x, I have not had time to go through and test all minor versions of the package because it requires build and deployment for every version test which takes times.
- package versions in play here: 
1. Azure-core: 1.30.0
2. azure-security-keyvault-secrets: 4.2.6

It's hard for me to provide more details because I have limited control over our runtime and its an enterprise repo. Please let me know if there is other info that can help resolve this.

**Comment by bennydean** [comment]
@gukoff Just an update, this seems to exist in azure-core 1.19. Somewhere between 1.15 and 1.19 inclusively, this bug shows up.

**Comment by vcolin7** [comment]
@bennydean, @gukoff Thank you for sharing your logs and what you've experienced so far. I think @anuchandy and I have a lead to where this might be happening in the stack and I will be looking into this next week.

**Comment by vcolin7** [comment]
I just had a chance to read your latest comment @gukoff. You are right to think `AccessTokenCache.getToken()` is the main suspect here. We make use of Reactor's `Sinks` class in some places to avoid having multiple of Reactor threads attempt to do the same work simultaneously, like retrieving an access token from the service. However, I think it's possible that if the first thread (the one actually doing the work) takes an unusually long time to complete, the rest will be caught in an async-busy-loop while waiting for anything other than `Mono.empty()` to be emitted, which in turn can cause the CPU use cycles on things like context switching, using it most of it up.

@anuchandy tested this on a small Docker container and we noticed CPU usage go up to 100%, whereas on my local machine with a lot more RAM and physical cores it used a good number of resources but never completely hogged the CPU, so it makes sense we see reports from Kubernetes users about it, for example.

A possible workaround for the time being ([other than the one mentioned above](https://github.com/Azure/azure-sdk-for-java/issues/28364#issuecomment-1114035758)) could be to avoid sharing a single client among many threads and instead creating one per thread, which isn't pretty but would ensure there is no more than one thread waiting for token retrieval per client.

I'm working on a fix at the moment and would love to have it in for the next release in a couple weeks, so once that's in I'll update this issue.

**Comment by vcolin7** [comment]
Actually, in your particular case @gukoff, I think it's related to `PollerFlux`. Like you said, it's likely it comes down to having multiple threads attempting to do the same thing while calling `Mono::repeatWhenEmpty`.

**Comment by vcolin7** [comment]
We've merged a PR with a fix for this issue when it involves sharing an `AccessTokenCache`. Also, after reviewing your logs one more time, I think was the issue in your case @gukoff. Sorry if I mislead you early on when I mentioned `PollerFlux`.

**Comment by ajdasari1231** [comment]
> We've merged a PR with a fix for this issue when it involves sharing an `AccessTokenCache`. Also, after reviewing your logs one more time, I think was the issue in your case @gukoff. Sorry if I mislead you early on when I mentioned `PollerFlux`.
> 
> The fix should be released in a couple of weeks.

Hello @vcolin7  Is this issue resolved?

**Comment by gukoff** [comment]
@vcolin7 curious, does [this change](https://github.com/Azure/azure-sdk-for-java/pull/31110) only fix the CPU consumption in the busy loop?

Because the more important issue was that the call to the keyvault didn't respect the timeout policy and hung forever.
Expected behaviour would be an exception after the given timeout, letting the user of the library retry or fall back.

**Comment by vcolin7** [comment]
Hi @werder06, I just wanted to check if the issue you were facing was fixed with the code changes we made a few months back. Have you had a chance to run your code using the latest version of the library?

**Comment by vcolin7** [comment]
@gukoff It's possible that timeouts were not honored due to thread starvation, but it's worth investigating, although I would treat it as a separate issue. Would you mind opening one and providing the details of what you saw with your tests? Thanks in advance :)

**Comment by vcolin7** [comment]
A small update: we're working on reproducing this issue on a Kubernetes pod but have not been able to do it successfully, we will communicate again with the customer reporting this and, if it cannot be reproduced with their setup, this issue will be closed.

**Comment by bennydean** [comment]
@vcolin7 I have tried the ExecutorService and it did not fix the issue. Any updates on this after we met and I provided details to the SR engineers on your team?

**Comment by vcolin7** [comment]
Closing this after @bennydean found the root cause of the issue to be related to his proxy configuration.
